onnx/checker.cc

```diff
@@ -228,6 +228,106 @@ void check_tensor(const TensorProto& tensor, const CheckerContext& ctx) {
 #undef check_field
 }
 
+void check_sequence(const SequenceProto& sequence, const CheckerContext& ctx) {
+  enforce_has_field(sequence, elem_type);
+  if (sequence.elem_type() == SequenceProto::TENSOR) {
+    for (const TensorProto& tensor : sequence.tensor_values()) {
+      check_tensor(tensor, ctx);
+    }
+  }
+  else if (sequence.elem_type() == SequenceProto::SPARSE_TENSOR) {
+    for (const SparseTensorProto& sparse_tensor : sequence.sparse_tensor_values()) {
+      check_sparse_tensor(sparse_tensor, ctx);
+    }
+  }
+  else if (sequence.elem_type() == SequenceProto::SEQUENCE) {
+    for (const SequenceProto& seq : sequence.sequence_values()) {
+      check_sequence(seq, ctx);
+    }
+  }
+  else if (sequence.elem_type() == SequenceProto::MAP) {
+    for (const MapProto& map : sequence.map_values()) {
+      check_map(map, ctx);
+    }
+  } else {
+    fail_check(
+        "Sequence ( Structure name: ",
+        sequence.name(),
+        ", elem_type: ",
+        sequence.elem_type(),
+        ") is not have a valid element type.");
+  }
+}
+
+void check_map(const MapProto& map, const CheckerContext& ctx) {
+  enforce_has_field(map, key_type);
+  if (map.key_type() == TensorProto::UNDEFINED) {
+    fail_check(
+        "setting key_type field (map name: ",
+        map.name(),
+        ") to UNDEFINED is not allowed");
+  }
+  // Check if key is a valid type, specifically INT8, INT16, INT32, INT64,
+  // UINT8, UINT16, UINT32, UINT64, or STRING.
+  if ((map.key_type() == TensorProto::FLOAT) ||
+      (map.key_type() == TensorProto::BOOL) ||
+      (map.key_type() == TensorProto::FLOAT16) ||
+      (map.key_type() == TensorProto::COMPLEX64) ||
+      (map.key_type() == TensorProto::COMPLEX128)) {
+    fail_check(
+        "setting key_type field (map name: ",
+        map.name(),
+        ") to invalid TensorProto key_type ",
+        map.key_type(),
+        " is not allowed");
+  }
+
+  enforce_has_field(map, value_type);
+  if (map.value_type() == SequenceProto::UNDEFINED) {
+    fail_check(
+        "setting value_type field (map name: ",
+        map.name(),
+        ") to UNDEFINED is not allowed");
+  }
+
+
+  #define check_repeated_field(field, type) \
+    if (map.field##_size() > 0) {          \
+      ++used_fields;                        \
+      check_type(type);                     \
+    }
+
+    check_repeated_field(keys, TensorProto::INT64);
+    check_repeated_field(string_keys, TensorProto::STRING);
+
+  #undef check_repeated_field
+
+    // Normally, used_fields is expected to be 1.
+    if (used_fields > 1) {
+      fail_check(
+          "Map (name: ",
+          map.name(),
+          ") should not contain more than one keys field.");
+    }
+
+  int num_keys = 0;
+  if (!map.has_keys()) {
+    num_keys = map.keys_size();
+  } else if (!map.has_string_keys()) {
+    num_keys = map.string_keys_size();
+  }
+
+  // if map.key_type == tensor -> tensor_values_size()
```

Comment:
nit: remove comments

Reply:
resolved with new fixes in latest update!